### PR TITLE
YARN images on Kubernetes.

### PR DESCRIPTION
**Pull request by k82cn**
@k8s-oncall

### CODE REVIEW

**Comment by timothysc** [comment]
@k82cn is there a reason why we wouldn't try to get the Dockerfiles into upstream hadoop and just pull from docker hub? and let the different vendors/projects produce?

**Comment by chrislovecnm** [comment]
cc me, we need to figure out where to put this. 

This will require a bunch more documentation ...

**Comment by chrislovecnm** [comment]
On mobile right now, let me figure out when I can review

**Comment by chrislovecnm** [comment]
@bgrant0607 we have talked in circles about were cool stuff like this should go. Who can make a decision where this should live?

This is needs some polish before we would merge, but how do we get it into the correct repo. Is this the correct home??

**Comment by davidopp** [comment]
cc/ @danisla @mblinn

**Comment by timothysc** [comment]
imho unless you place it where a given upstream publishes it's artifacts you will always bit-rot and potentially have a bad user experience b/c it's questionable what level of vetting has occurred, and what versions are supported.

/cc @c0s

**Comment by bgrant0607** [comment]
@chrislovecnm 

@timothysc is right that upstream is ideal. With respect to our repos, though, this is probably the most appropriate. Where will the Kubernetes manifest files live?

**Comment by bgrant0607** [comment]
cc @kubernetes/examples

**Comment by davidopp** [comment]
I kinda agree that using an image somebody else maintains would be better. One of the other folks in the community who has been working with YARN on Kubernetes has been using the public Docker images from SequenceIQ. (I didn't investigate these images, but that's what they said.)

**Comment by k82cn** [comment]
@davidopp , it seems SequenceIQ/hadoop deploy hadoop in one docker; but this patch will start YARN master & YARN agent  in different docker container.

@kubernetes/examples , I'd like also to create YML files to deploy a YARN cluster as service, so where should I put those Dockerfiles?

**Comment by danisla** [comment]
@davidopp, @k82cn this looks like a pretty similar approach to how I've been running YARN on K8S, I was just customizing the SequenceIQ image at runtime to break apart the processes to run in separate containers and avoid building and maintaining my own image. 

The SequenceIQ guys documented the [hadoop native build process](https://github.com/sequenceiq/docker-hadoop-build/blob/master/Dockerfile) This lets users customize their distribution (like adding protobuf) in a portable way. It would be nice if users had that option to customize in this repo as a separate (but lengthy) make action.

I think this is a good start for a hadoop image, just have a few questions:

- Is there anyway to get rid of the ssh dependency and avoid committing private keys?
- Given that the hostname of the pod is being added via template, how does service discovery work? Are you using a PetSet, some specific networking or DNS requirement?

Also, I agree that the kubernetes manifest files should go in [kubernetes/examples](https://github.com/kubernetes/kubernetes/tree/master/examples)

**Comment by timothysc** [comment]
> I'd like also to create YML files to deploy a YARN cluster as service, so where should I put those Dockerfile . 

The Dockerfiles should try to get pushed upstream where possible.  I can help poke the hadoop contacts that I know.  /cc @steveloughran @wattsteve 

re:manifest files for kube? I no longer know, as we've fragmented the repo.  IMHO it's a deployment.

**Comment by k82cn** [comment]
@timothysc , thanks very much; I'd like to work with Hadoop to upstream, would you start a email thread or send hadoop contacts to me?

@davidopp, @timothysc, and where should we put images? Docker hub in my personal account?

After the image & Dockerfile done, I'll update kubenetes/example/yarn accordingly.

**Comment by k82cn** [comment]
@danisla 

> * Is there anyway to get rid of the ssh dependency and avoid committing private keys?

Good point, I'll find a way to avoid it.

> * Given that the hostname of the pod is being added via template, how does service discovery work? Are you using a PetSet, some specific networking or DNS requirement?

No PetSet; I update yarn's configuration to use ip for communication; only yarn master is a Service.

**Comment by davidopp** [comment]
> where should we put images? Docker hub in my personal account?

I think this repo is the right place to put images if we can't get someone else to maintain them. As for related files like the .yaml's for bringing up a cluster, I'm not sure whether those would go in kubernetes/examples/ or a new repo under kubernetes/ -- maybe @bgrant0607 can comment.

**Comment by wattsteve** [comment]
We're starting to have a convention in kubernetes/examples/ where we have a
subdirs that contain further examples relating to a specific functional
area. For examples kubernetes/examples/volumes or
kubernetes/examples/storage. I think we might consider a
kubernetes/examples/petsets where we can store YAML examples of various
petset impls. I think this would be useful in that we can build testing
against these into regression tests for the petset framework as it
continues to evolve.

**Comment by chrislovecnm** [comment]
@wattsteve I would prefer to keep petset yaml with the examples that are using them in particular.

@k82cn can we use upstream images?

**Comment by davidopp** [comment]
Yeah this isn't really a PetSet example -- it's an application framework that might happen to use one or more Kubernetes features like PetSet.

**Comment by k82cn** [comment]
@chrislovecnm , which upstream images do you mean? If there's any place I can contribute, I'd like to pull it there.

**Comment by bgrant0607** [comment]
I would love to move kubernetes/examples to another repo, but work would be required in order to test them (validation unit tests and e2e).

cc @kubernetes/examples

**Comment by chrislovecnm** [comment]
@bgrant0607 what is your take where we should put this example?  contrib?

File: yarn/Dockerfile.agent [hunk 1]
```diff
@@ -0,0 +1,29 @@
+FROM ubuntu:14.04
```
Comment (chrislovecnm) [inline yarn/Dockerfile.agent]: This should be up to date ... it is note.

File: yarn/Dockerfile.agent [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+MAINTAINER Klaus Ma <madaxa@cn.ibm.com>
+
+RUN apt-get update && apt-get -y install openjdk-7-jdk openssh-server
```
Comment (chrislovecnm) [inline yarn/Dockerfile.agent]: can we run jdk-8 please?  7 is nasty slow.

File: yarn/Makefile [hunk 3]
```diff
@@ -0,0 +1,31 @@
+all: yarn-agent yarn-master
+push: push-agent push-master
+.PHONY: push push-agent push-master yarn-agent yarn-master
+
+TAG = 2.7.2
+
+init:
+	if [ ! -e "hadoop-2.7.2.tar.gz" ] ; then  wget http://apache.fayea.com/hadoop/common/hadoop-2.7.2/hadoop-2.7.2.tar.gz; fi
```
Comment (chrislovecnm) [inline yarn/Makefile]: Can we parameterize this?

File: yarn/.ssh/id_rsa [hunk 4]
```diff
@@ -0,0 +1,27 @@
+-----BEGIN RSA PRIVATE KEY-----
```
Comment (chrislovecnm) [inline yarn/.ssh/id_rsa]: Can we generate the id_rsa and authorized keys in the Makefile?  They should be different between environments.

File: yarn/Dockerfile.agent [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+COPY hadoop-2.7.2.tar.gz /opt/
+
+RUN cd /opt/ && tar zxvf hadoop-2.7.2.tar.gz && rm hadoop-2.7.2.tar.gz
+
+COPY ./agent_bootstrap.sh /opt/bootstrap.sh
```
Comment (chrislovecnm) [inline yarn/Dockerfile.agent]: each of these COPY's produce a layer.  Can we get them in one COPY.  Pattern I usually follow is copy the files in then use RUN to move them to the correct place.  This docker is creating a ton of layers.

File: yarn/README.md [hunk 6]
```diff
@@ -0,0 +1,6 @@
+# YARN
```
Comment (chrislovecnm) [inline yarn/README.md]: In order to get this in, wherever we are able to put it.  We need considerable more documentation.  Look at the examples in the k8s project.  Step by step.

File: yarn/master_bootstrap.sh [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+sed "s/__HOSTNAME__/"$(hostname)"/g" /root/yarn-site.xml.template   > $HADOOP_CONF_DIR/yarn-site.xml
+sed "s/__HOSTNAME__/"$(hostname)"/g" /root/hdfs-site.xml.template   > $HADOOP_CONF_DIR/hdfs-site.xml
+sed "s/__HOSTNAME__/"$(hostname)"/g" /root/mapred-site.xml.template > $HADOOP_CONF_DIR/mapred-site.xml
+
+mkdir /opt/hadoop272/dfs/name -p
+mkdir /opt/hadoop272/dfs/data -p
+
+$HADOOP_PREFIX/bin/hdfs namenode -format yarn_272_hdfs
+
+$HADOOP_PREFIX/sbin/hadoop-daemon.sh --config $HADOOP_CONF_DIR --script hdfs start namenode
```
Comment (chrislovecnm) [inline yarn/master_bootstrap.sh]: Can we use supervisor for these?  Can these be run in separate attached containers, or do we need multiple processes running inside the same docker?

**Comment by chrislovecnm** [comment]
@k82cn if we are able to use officially maintained dockers and make them k8s friendly that would be the best.  I am not certain if maintained dockers exist or not.  Lots of comments for yah.

**Comment by k82cn** [comment]
@chrislovecnm , let me also send an email to YARN dev@/usr@ for dockers.

**Comment by k82cn** [comment]
@chrislovecnm ,  there's no official docker image in Hadoop after checking in user@hadoop.

**Comment by davidopp** [comment]
So it seems the SequenceIQ Docker image is the closest to "official"?
https://github.com/sequenceiq/hadoop-docker

**Comment by k82cn** [comment]
@davidopp , yes, it's closed to "official"; but it need to configure manually :(. It can not deploy by k8s.

**Comment by k82cn** [comment]
@chrislovecnm @davidopp , there's not official images here, but ozawa@apache.org will start a thread in Hadoop community to discuss it. I'll follow up with them and keep you posted.

**Comment by k82cn** [comment]
Create a JIRA in Hadoop community to trace the Docker images at https://issues.apache.org/jira/browse/HADOOP-13397

Keep you posted if any news :).

**Pull request closed without merging**